spec/controllers/assignments_controller_spec.rb

```diff
@@ -378,4 +405,105 @@
       end
     end
   end
+
+  describe 'build complete tree' do
+    describe 'when fetching directory contents is successful' do
+      let(:directory_contents) { [ 'file1.cpp', 'file2.cpp' ] }
+
+      before do
+        allow(controller).to receive(:fetch_directory_contents).and_return(directory_contents)
+      end
+
+      it 'fetches directory contents successfully' do
+        result = controller.send(:build_complete_tree, assignment)
+        expect(result).to eq(directory_contents)
+      end
+    end
+
+    context 'when there is a GitHub API error' do
+      before do
+        allow(controller).to receive(:fetch_directory_contents).and_raise(Octokit::Error.new(response_status: 500, message: 'Some error occurred'))
+        allow(Rails.logger).to receive(:error)
+      end
+
+      it 'logs the error and returns an empty array' do
+        result = controller.send(:build_complete_tree, assignment)
+
+        expect(Rails.logger).to have_received(:error) do |message|
+          expect(message).to start_with("GitHub API Error:")
+        end
+        expect(result).to eq([])
+      end
+    end
+  end
+
+  describe 'fetch directory contents' do
+    let(:client) { instance_double('Client') }
+    let(:repo) { 'example_repo' }
+    let(:path) { 'root' }
+
+    let(:file_item) do
+      double(
+        name: 'file1.txt',
+        path: 'root/file1.txt',
+        type: 'file'
+      )
+    end
+
+    let(:dir_item) do
+      double(
+        name: 'dir1',
+        path: 'root/dir1',
+        type: 'dir'
+      )
+    end
+
+    let(:nested_file_item) do
+      double(
+        name: 'file2.txt',
+        path: 'root/dir1/file2.txt',
+        type: 'file'
+      )
+    end
+
+    before do
+      allow(client).to receive(:contents).with(repo, path: path).and_return([ file_item, dir_item ])
+      allow(client).to receive(:contents).with(repo, path: 'root/dir1').and_return([ nested_file_item ])
+    end
+
+    it 'returns directory contents with nested structure for directories' do
+      expected_result = [
+        {
+          name: 'file1.txt',
+          path: 'root/file1.txt',
+          type: 'file'
+        },
+        {
+          name: 'dir1',
+          path: 'root/dir1',
+          type: 'dir',
+          children: [
+            {
+              name: 'file2.txt',
+              path: 'root/dir1/file2.txt',
+              type: 'file'
+            }
+          ]
+        }
+      ]
+
+      result = controller.send(:fetch_directory_contents, client, repo, path)
+      expect(result).to eq(expected_result)
+    end
+
+    it 'encodes spaces in directory paths with ++' do
+      allow(dir_item).to receive(:path).and_return('root/dir 1') # Path with space
```

Comment:
Instead of using `root/dir1` as the directory path, I think it'd be good to use `tests/c++` instead. Might give the person going through the test code, some context as to why the `gsub` operation was necessary.